app/script/conversation/ConversationCellState.js

```diff
@@ -23,279 +23,282 @@ window.z = window.z || {};
 window.z.conversation = z.conversation || {};
 
 z.conversation.ConversationCellState = (() => {
-  function is_alert(message_et) {
-    return message_et.is_ping() || (message_et.is_call() && message_et.was_missed());
-  }
+  const ACTIVITY_TYPE = {
+    CALL: 'ConversationCellState.ACTIVITY_TYPE.CALL',
+    MESSAGE: 'ConversationCellState.ACTIVITY_TYPE.MESSAGE',
+    PING: 'ConversationCellState.ACTIVITY_TYPE.PING',
+  };
 
-  function generate_activity_string(activities) {
-    const activity_strings = [];
+  const _accumulateActivity = conversationEntity => {
+    const unreadEvents = conversationEntity.unread_events();
+    const activities = {
+      [ACTIVITY_TYPE.CALL]: 0,
+      [ACTIVITY_TYPE.MESSAGE]: 0,
+      [ACTIVITY_TYPE.PING]: 0,
+    };
 
-    for (const activity in activities) {
-      if (activities.hasOwnProperty(activity)) {
-        const count = activities[activity];
-
-        switch (activity) {
-          case 'message':
-            if (count === 1) {
-              activity_strings.push(z.l10n.text(z.string.conversationsSecondaryLineNewMessage, count));
-            } else if (count > 1) {
-              activity_strings.push(z.l10n.text(z.string.conversationsSecondaryLineNewMessages, count));
-            }
-            break;
-          case 'ping':
-            if (count === 1) {
-              activity_strings.push(z.l10n.text(z.string.conversationsSecondaryLinePing, count));
-            } else if (count > 1) {
-              activity_strings.push(z.l10n.text(z.string.conversationsSecondaryLinePings, count));
-            }
-            break;
-          case 'call':
-            if (count === 1) {
-              activity_strings.push(z.l10n.text(z.string.conversationsSecondaryLineMissedCall, count));
-            } else if (count > 1) {
-              activity_strings.push(z.l10n.text(z.string.conversationsSecondaryLineMissedCalls, count));
-            }
-            break;
-          default:
-        }
+    for (const messageEntity of unreadEvents) {
+      if (messageEntity.is_call() && messageEntity.was_missed()) {
+        activities[ACTIVITY_TYPE.CALL] = activities[ACTIVITY_TYPE.CALL] + 1;
+      } else if (messageEntity.is_ping()) {
+        activities[ACTIVITY_TYPE.PING] = [ACTIVITY_TYPE.PING] + 1;
+      } else if (messageEntity.is_content()) {
+        activities[ACTIVITY_TYPE.MESSAGE] = activities[ACTIVITY_TYPE.MESSAGE] + 1;
       }
     }
 
-    return activity_strings.join(', ');
-  }
+    return _generateActivityString(activities);
+  };
 
-  function accumulate_activity(conversation_et) {
-    const unread_events = conversation_et.unread_events();
-    const activities = {
-      call: 0,
-      message: 0,
-      ping: 0,
-    };
+  const _generateActivityString = activities => {
+    const activityStrings = [];
+
+    for (const activity in activities) {
```

Comment:
To avoid ```hasOwnProperty()``` use ```Object.keys(activities).map()```, which would make it more expressive and therefore more readable.